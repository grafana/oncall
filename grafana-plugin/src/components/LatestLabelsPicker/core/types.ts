export interface ItemRepresentation {
  id: string;
  repr: string;
}

export interface ItemGroup {
  key: ItemRepresentation;
  values: ItemRepresentation[];
}

export interface ItemSelected {
  key: ItemRepresentation;
  value: ItemRepresentation;
}