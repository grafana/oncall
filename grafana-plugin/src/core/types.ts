export interface ItemGroup {
  key: any;
  values: any[];
}

export interface ItemSelected {
  key: any;
  value: any;
}

export enum LabelInputType {
  key = 'key',
  value = 'value',
}

export interface ServiceLabelValidator {
  isValid: boolean;
  errorMessage?: string;
}
