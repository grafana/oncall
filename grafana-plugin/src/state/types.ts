import { AppPluginMeta, KeyValue } from '@grafana/data';

import { RootStore } from 'state/rootStore';

export interface WithStoreProps {
  store: RootStore;
}

export interface PageProps<T extends KeyValue = KeyValue> {
  meta: AppPluginMeta<T>;
  query: KeyValue;
}

export interface SelectOption {
  value: string | number;
  display_name: string;
}
