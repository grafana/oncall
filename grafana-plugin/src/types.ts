import { AppRootProps as BaseAppRootProps, AppPluginMeta, CurrentUserDTO, PluginConfigPageProps } from '@grafana/data';

export type OnCallPluginMetaJSONData = {
  stackId: number;
  orgId: number;
  onCallApiUrl: string;
  insightsDatasource?: string;
  license: string;
};

export type OnCallPluginMetaSecureJSONData = {
  grafanaToken: string;
  onCallApiToken: string;
};

export type AppRootProps = BaseAppRootProps<OnCallPluginMetaJSONData>;

// NOTE: it is possible that plugin.meta.jsonData is null (ex. on first-ever setup)
// the typing on AppPluginMeta does not seem correct atm..
export type OnCallAppPluginMeta = AppPluginMeta<null | OnCallPluginMetaJSONData>;
export type OnCallPluginConfigPageProps = PluginConfigPageProps<OnCallAppPluginMeta>;

// Extension points that other plugins can use to hook into the OnCall app.
export enum OnCallPluginExtensionPoints {
  AlertGroupAction = 'plugins/grafana-oncall-app/alert-group/action',
}

declare global {
  export interface Window {
    // https://github.com/grafana/grafana/blob/78bef7a26a799209b5307d6bde8e25fcb4fbde7d/public/views/index-template.html#L251-L258
    grafanaBootData: {
      user: CurrentUserDTO;
      settings: {
        unifiedAlertingEnabled: boolean;
        unifiedAlerting: { minInterval: string };
      };
    };
    RECAPTCHA_SITE_KEY: string;
    grecaptcha: any;
    dataLayer: any;
    mixpanel: any;
  }
}
