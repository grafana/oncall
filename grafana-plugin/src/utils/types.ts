import { RenderedCell } from 'rc-table/lib/interface';

export interface TableColumn {
  width?: string | number;
  title: string;
  key: string;
  render: (value: any, record: any, index: number) => React.ReactNode | RenderedCell<any>;
  className?: string;
  grow?: number;
}

export type PropertiesThatExtendsAnotherClass<OriginalObj, AnotherClass> = keyof {
  [Prop in keyof OriginalObj as OriginalObj[Prop] extends AnotherClass ? Prop : never]: unknown;
};

// IfEquals, WritableKeys, ReadonlyKeys based on https://stackoverflow.com/a/49579497/4931398
export type IfEquals<X, Y, A = X, B = never> = (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2
  ? A
  : B;

export type WritableKeys<T> = {
  [P in keyof T]-?: IfEquals<{ [Q in P]: T[P] }, { -readonly [Q in P]: T[P] }, P>;
}[keyof T];

export type ReadonlyKeys<T> = {
  [P in keyof T]-?: IfEquals<{ [Q in P]: T[P] }, { -readonly [Q in P]: T[P] }, never, P>;
}[keyof T];

export type OmitReadonlyMembers<T> = Omit<T, ReadonlyKeys<T>>;
