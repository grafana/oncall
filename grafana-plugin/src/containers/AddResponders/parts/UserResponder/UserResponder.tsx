import React, { FC } from 'react';

import { SelectableValue } from '@grafana/data';
import { ActionMeta, HorizontalGroup, IconButton } from '@grafana/ui';
import cn from 'classnames/bind';

import { Avatar } from 'components/Avatar/Avatar';
import { Text } from 'components/Text/Text';
import styles from 'containers/AddResponders/AddResponders.module.scss';
import { UserResponder as UserResponderType } from 'containers/AddResponders/AddResponders.types';
import { NotificationPoliciesSelect } from 'containers/AddResponders/parts/NotificationPoliciesSelect/NotificationPoliciesSelect';

const cx = cn.bind(styles);

type Props = UserResponderType & {
  onImportantChange: (value: SelectableValue<number>, actionMeta: ActionMeta) => void | {};
  handleDelete: React.MouseEventHandler<HTMLButtonElement>;
  disableNotificationPolicySelect?: boolean;
};

export const UserResponder: FC<Props> = ({
  important,
  data: { avatar, username },
  onImportantChange,
  handleDelete,
  disableNotificationPolicySelect = false,
}) => (
  <li>
    <HorizontalGroup justify="space-between">
      <HorizontalGroup>
        <div className={cx('timeline-icon-background', { 'timeline-icon-background--green': true })}>
          <Avatar size="medium" src={avatar} />
        </div>
        <Text className={cx('responder-name')}>{username}</Text>
      </HorizontalGroup>
      <HorizontalGroup>
        <NotificationPoliciesSelect
          disabled={disableNotificationPolicySelect}
          important={important}
          onChange={onImportantChange}
        />
        <IconButton
          data-testid="user-responder-delete-icon"
          tooltip="Remove responder"
          name="trash-alt"
          onClick={handleDelete}
        />
      </HorizontalGroup>
    </HorizontalGroup>
  </li>
);
