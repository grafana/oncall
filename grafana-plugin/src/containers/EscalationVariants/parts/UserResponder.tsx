import React, { FC } from 'react';

import { HorizontalGroup, Icon, Select, IconButton, Tooltip } from '@grafana/ui';
import cn from 'classnames/bind';

import Avatar from 'components/Avatar/Avatar';
import PluginLink from 'components/PluginLink/PluginLink';
import Text from 'components/Text/Text';
import styles from 'containers/EscalationVariants/EscalationVariants.module.scss';
import {
  ResponderBaseProps,
  UserResponder as UserResponderType,
} from 'containers/EscalationVariants/EscalationVariants.types';

const cx = cn.bind(styles);

type UserResponderProps = ResponderBaseProps & Pick<UserResponderType, 'important' | 'data'>;

const UserResponder: FC<UserResponderProps> = ({ important, data, onImportantChange, handleDelete }) => (
  <li>
    <HorizontalGroup justify="space-between">
      <HorizontalGroup>
        <div className={cx('timeline-icon-background', { 'timeline-icon-background--green': true })}>
          <Avatar size="medium" src={data?.avatar} />
        </div>
        <Text className={cx('responder-name')}>{data?.username}</Text>
        {data.notification_chain_verbal.default || data.notification_chain_verbal.important ? (
          <HorizontalGroup>
            <Text type="secondary">by</Text>
            <Select
              className={cx('select')}
              width="auto"
              isSearchable={false}
              value={Number(important)}
              options={[
                {
                  value: 0,
                  label: 'Default',
                  description: 'Use "Default notifications" from user\'s personal settings',
                },
                {
                  value: 1,
                  label: 'Important',
                  description: 'Use "Important notifications" from user\'s personal settings',
                },
              ]}
              // @ts-ignore
              isOptionDisabled={({ value }) =>
                (value === 0 && !data.notification_chain_verbal.default) ||
                (value === 1 && !data.notification_chain_verbal.important)
              }
              getOptionLabel={({ value, label }) => {
                return (
                  <Text
                    type={
                      (value === 0 && !data.notification_chain_verbal.default) ||
                      (value === 1 && !data.notification_chain_verbal.important)
                        ? 'disabled'
                        : 'primary'
                    }
                  >
                    {label}
                  </Text>
                );
              }}
              onChange={onImportantChange}
            />
            <Text type="secondary">notification policies</Text>
          </HorizontalGroup>
        ) : (
          <HorizontalGroup>
            <Tooltip content="User doesn't have configured notification policies">
              <Icon name="exclamation-triangle" style={{ color: 'var(--error-text-color)' }} />
            </Tooltip>
          </HorizontalGroup>
        )}
      </HorizontalGroup>
      <HorizontalGroup>
        <PluginLink className={cx('hover-button')} target="_blank" query={{ page: 'users', id: data.pk }}>
          <IconButton
            tooltip="Open user profile in new tab"
            style={{ color: 'var(--always-gray)' }}
            name="external-link-alt"
          />
        </PluginLink>
        <IconButton tooltip="Remove responder" className={cx('hover-button')} name="trash-alt" onClick={handleDelete} />
      </HorizontalGroup>
    </HorizontalGroup>
  </li>
);

export default UserResponder;
