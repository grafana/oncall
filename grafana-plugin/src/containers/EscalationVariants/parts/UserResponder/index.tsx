import React, { FC } from 'react';

import { HorizontalGroup, IconButton } from '@grafana/ui';
import cn from 'classnames/bind';

import Avatar from 'components/Avatar/Avatar';
import Text from 'components/Text/Text';
import styles from 'containers/EscalationVariants/EscalationVariants.module.scss';
import {
  ResponderBaseProps,
  UserResponder as UserResponderType,
} from 'containers/EscalationVariants/EscalationVariants.types';
import NotificationPoliciesSelect from 'containers/EscalationVariants/parts/NotificationPoliciesSelect';

const cx = cn.bind(styles);

type UserResponderProps = ResponderBaseProps &
  Pick<UserResponderType, 'important' | 'data'> & {
    disableNotificationPolicySelect?: boolean;
  };

const UserResponder: FC<UserResponderProps> = ({
  important,
  data: { avatar, username },
  onImportantChange,
  handleDelete,
  disableNotificationPolicySelect = false,
}) => (
  <li>
    <HorizontalGroup justify="space-between">
      <HorizontalGroup>
        <div className={cx('timeline-icon-background', { 'timeline-icon-background--green': true })}>
          <Avatar size="medium" src={avatar} />
        </div>
        <Text className={cx('responder-name')}>{username}</Text>
      </HorizontalGroup>
      <HorizontalGroup>
        <NotificationPoliciesSelect
          disabled={disableNotificationPolicySelect}
          important={important}
          onChange={onImportantChange}
        />
        <IconButton
          data-testid="user-responder-delete-icon"
          tooltip="Remove responder"
          name="trash-alt"
          onClick={handleDelete}
        />
      </HorizontalGroup>
    </HorizontalGroup>
  </li>
);

export default UserResponder;
