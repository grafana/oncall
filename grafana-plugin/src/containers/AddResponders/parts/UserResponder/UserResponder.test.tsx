import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ApiSchemas } from 'network/oncall-api/api.types';

import { UserResponder } from './UserResponder';

describe('UserResponder', () => {
  const user = {
    avatar: 'http://avatar.com/',
    username: 'johnsmith',
  } as ApiSchemas['UserIsCurrentlyOnCall'];

  test('it renders data properly', () => {
    const component = render(
      <UserResponder important data={user} onImportantChange={() => {}} handleDelete={() => {}} />
    );
    expect(component.container).toMatchSnapshot();
  });

  test('it calls the delete callback', async () => {
    const handleDelete = jest.fn();

    render(<UserResponder important data={user} onImportantChange={() => {}} handleDelete={handleDelete} />);

    const deleteIcon = await screen.findByTestId('user-responder-delete-icon');
    await userEvent.click(deleteIcon);

    expect(handleDelete).toHaveBeenCalledTimes(1);
  });
});
