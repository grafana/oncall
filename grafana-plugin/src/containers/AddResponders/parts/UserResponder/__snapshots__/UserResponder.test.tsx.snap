// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserResponder it renders data properly 1`] = `
<div>
  <li>
    <div
      class="css-1mhys9y-horizontal-group"
      style="width: 100%; height: 100%;"
    >
      <div
        class="css-18qv8yz-layoutChildrenWrapper"
      >
        <div
          class="css-ffyaiw-horizontal-group"
          style="width: 100%; height: 100%;"
        >
          <div
            class="css-18qv8yz-layoutChildrenWrapper"
          >
            <div
              class="timeline-icon-background timeline-icon-background--green"
            >
              <img
                class="root avatarSize-medium"
                data-testid="test__avatar"
                src="http://avatar.com/"
              />
            </div>
          </div>
          <div
            class="css-18qv8yz-layoutChildrenWrapper"
          >
            <span
              class="root text text--undefined text--medium responder-name"
            >
              johnsmith
            </span>
          </div>
        </div>
      </div>
      <div
        class="css-18qv8yz-layoutChildrenWrapper"
      >
        <div
          class="css-ffyaiw-horizontal-group"
          style="width: 100%; height: 100%;"
        >
          <div
            class="css-18qv8yz-layoutChildrenWrapper"
          >
            <div
              class="css-1nmqu8c-input-wrapper select css-8k5qe3-SelectContainer"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class="css-1i88p6p"
              >
                <div
                  class="css-1q0c0d5-grafana-select-value-container"
                >
                  <div
                    class="css-8nwx1l-singleValue css-upz218-SingleValue"
                  >
                    Important
                  </div>
                  <input
                    aria-autocomplete="list"
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-readonly="true"
                    class="css-mohuvp-dummyInput-DummyInput"
                    id="react-select-2-input"
                    inputmode="none"
                    role="combobox"
                    tabindex="0"
                    value=""
                  />
                </div>
                <div
                  class="css-zyjsuv-input-suffix"
                >
                  <div
                    class="css-1j2891d-Icon"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-18qv8yz-layoutChildrenWrapper"
          >
            <button
              aria-label="Remove responder"
              class="css-17584xm"
              data-testid="user-responder-delete-icon"
              tabindex="0"
              type="button"
            >
              <div
                class="css-1j2891d-Icon"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </li>
</div>
`;
